### PR TITLE
[Gradient Compression] Make FP16 compression as a wrapper that can be combined with other communication hooks

### PR DESCRIPTION
**Pull request by iseessel**
Create a FP16 wrapper that can combine FP16 gradient compression with any gradient compression algorithm. 

Differential Revision: D26978832

Fixes #{https://github.com/pytorch/pytorch/issues/52986}

### CODE REVIEW

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 1]
```diff
@@ -65,3 +66,33 @@ def decompress(fut):
         return [decompressed_tensor]
 
     return fut.then(decompress)
+
+def fp16_compress_wrapper(hook: Callable[..., torch.futures.Future]) -> Callable[..., torch.futures.Future]:
```
Comment (SciPioneer) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:70]: Can you use `Callable[[Any, dist.GradBucket], torch.futures.Future]` as the hook type instead?

File: test/distributed/test_c10d.py [hunk 2]
```diff
@@ -3829,6 +3829,28 @@ def _test_default_ddp_comm_hooks_nccl(self, gradient_as_bucket_view=False):
...
             self._run_and_verify_hook(gpu_model, 8, 0.25 * torch.ones(2, 2))
 
+    def _test_fp16_compress_wrapper(self, gradient_as_bucket_view=False):
+        """
+        This unit test verifies whether wrapping the ALLREDUCE and POWER_SGD hooks with
+        the FP16_WRAPPER can give the same result as when there is no hook registered.
+        """
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+        power_sgd_state = powerSGD.PowerSGDState(process_group=process_group)
```
Comment (SciPioneer) [inline test/distributed/test_c10d.py:3839]: Nit: "power_sgd_state" -> "powerSGD_state" for naming consistency in this file.

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py (hunk 1 above)
Comment (iseessel) [reply torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:70]: Yep that's better

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 3]
```diff
@@ -65,3 +66,33 @@ def decompress(fut):
...
+    """
+    This wrapper casts the input gradient tensors of a given DDP communication hook to half-precision
+    floating point format (torch.float16), and casts the resulting tensors of the given hook back to
```
Comment (SciPioneer) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:73]: Nit: Please quote the data types with two grave accents (``).

``torch.float16``
``float32``

**Review by wayi1 — APPROVED** [review]
LGTM! Thanks for contributing to PyTorch community!

File: torch/lib/c10d/comm.hpp [hunk 4]
```diff
@@ -57,6 +57,11 @@ class GradBucket {
     return index_ == 0;
   }
 
+  // Overwrites tensors at a specific index.
```
Comment (iseessel) [inline torch/lib/c10d/comm.hpp:60]: I couldn't find test files for this or the init.cpp anywhere -- let me know if I missed it!

Comment (SciPioneer) [inline torch/lib/c10d/comm.hpp:60]: Nit: Overwrites **a tensor** at a specific index in the bucket.

Comment (SciPioneer) [reply torch/lib/c10d/comm.hpp:60]: There isn't any test for init.cpp, as all the C++ functions can be viewed as internal functions. We only need to test python public APIs.

File: torch/lib/c10d/comm.hpp [hunk 5]
```diff
@@ -57,6 +57,11 @@ class GradBucket {
...
+  void setTensor(at::Tensor& tensor, int i) {
+    tensors_[i] = tensor;
```
Comment (SciPioneer) [inline torch/lib/c10d/comm.hpp:62]: You need to use `tensors_[i].copy_(tensor)` to reduce the memory footprint.

File: torch/csrc/distributed/c10d/init.cpp [hunk 6]
```diff
@@ -249,6 +249,15 @@ to apply layer-wise operations.
...
     This also means that this bucket corresponds to the first few layers in the forward pass.
+)")
+      .def(
+          "set_tensor",
+          &::c10d::GradBucket::setTensor,
+          py::arg("tensor"),
+          py::arg("i"),
+          py::call_guard<py::gil_scoped_release>(),
+          R"(
+Replaces _tensor at index i with tensor input.
```
Comment (SciPioneer) [inline torch/csrc/distributed/c10d/init.cpp:260]: There is only "tensors_" member, no "_tensor" member in `GradBucket` class.

Nit: Replaces the ith tensor in the bucket with the input tensor.

**Comment by SciPioneer** [comment]
Thanks for the fix! It's almost done. 

Still need to update two files:
1) Add `def set_tensor(self, tensor: Tensor, i: int): ...`
at https://github.com/pytorch/pytorch/blob/master/torch/_C/_distributed_c10d.pyi#L23
Declaring the new python method is necessary for some platforms.

2) Add `.. autofunction:: torch.distributed.GradBucket.set_tensor`
after ".. autofunction:: torch.distributed.GradBucket.is_the_last_bucket_to_allreduce" in `ddp_comm_hooks.rst`
This will update the PyTorch tutorial at https://pytorch.org/docs/master/ddp_comm_hooks.html

**Comment by SciPioneer** [comment]
LGTM. Thanks for the help!
